### PR TITLE
Evm revert

### LINKED ISSUE #191: contracts tests: snapshot and reset before/after each test
We can use the [`evm_snapshot`](https://hardhat.org/hardhat-network/docs/reference#evm_snapshot) hardhat opcode in a `beforeEach` hook and a `evm_revert` call in an `afterEach` hook to give each test a clean environment to operate in.

### LINKED ISSUE #226: Cascading test failures
In the contract tests we use `evm_snapshot`/`evm_revert` to reset state after a test. We currently put the `evm_revert` call at the end of the test. This means that if the test fails (because e.g. an error is thrown) the `evm_revert` call is not executed. This causes the other tests to fail because the contract state is left dirty.

Related #223 #191

### PR DESCRIPTION
**Pull request by vivianjeng**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

closes #191
closes #226 

## What is the new behavior?

- A snapshot can only be reverted once: https://hardhat.org/hardhat-network/docs/reference#evm_snapshot
- So I use `beforeEach` to record all new snapshots

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/contracts/test/userStateTransition.test.ts [hunk 1]
```diff
@@ -634,11 +638,12 @@ describe('User State Transition', function () {
 
         const users = 3
         const epochs = 3
-        let _stateTree = {}
+        let stateTreeMap = {}
```
Comment (vimwitch) [inline packages/contracts/test/userStateTransition.test.ts:641]: Does this need to be a map or can it just be 2 separate variables?

File: packages/core/test/stateTree.test.ts [hunk 2]
```diff
@@ -17,21 +17,29 @@ describe('State tree', function () {
     this.timeout(30 * 60 * 1000)
 
     let unirepContract
-    let stateTree
+    let config
```
Comment (vimwitch) [inline packages/core/test/stateTree.test.ts:20]: This doesn't need to be a global variable.

File: packages/core/test/stateTree.test.ts [hunk 3]
```diff
@@ -137,6 +148,7 @@ describe('State tree', function () {
             .map(() => {
                 return new ZkIdentity()
             })
+        let stateTree = new IncrementalMerkleTree(config.stateTreeDepth)
```
Comment (vimwitch) [inline packages/core/test/stateTree.test.ts:151]: `const` unless the variable is being re-assigned.

**Review by vimwitch — CHANGES_REQUESTED** [review]

**Comment by vimwitch** [comment]
Opened #244 to help test with larger state trees

File: packages/core/test/stateTree.test.ts (hunk 3 above)
Comment (vivianjeng) [reply packages/core/test/stateTree.test.ts]: [5f4ff50](https://github.com/Unirep/Unirep/pull/243/commits/5f4ff505d0f5fc341c1f75c9e8cdcab1d6c82606)

File: packages/core/test/stateTree.test.ts (hunk 2 above)
Comment (vivianjeng) [reply packages/core/test/stateTree.test.ts]: [5f4ff50](https://github.com/Unirep/Unirep/pull/243/commits/5f4ff505d0f5fc341c1f75c9e8cdcab1d6c82606)

File: packages/contracts/test/userStateTransition.test.ts (hunk 1 above)
Comment (vivianjeng) [reply packages/contracts/test/userStateTransition.test.ts]: Fix it here [a3aa6ce](https://github.com/Unirep/Unirep/pull/243/commits/a3aa6ce14e3b3c9a75963f15f108b05c2dca3b4d)

**Review by vimwitch — APPROVED** [review]

**Pull request merged**